python/lsst/daf/butler/_butler.py

```diff
@@ -547,6 +547,62 @@ def _findDatasetRef(self, datasetRefOrType: Union[DatasetRef, DatasetType, str],
         else:
             idNumber = None
         timespan: Optional[Timespan] = None
+
+        # Process dimension records that are using record information
+        # rather than ids
+        newDataId: dict[Any, Any] = {}
+        byRecord: dict[Any, dict[str, Any]] = defaultdict(dict)
+        for k, v in dataId.items():
+            # If we have a Dimension we do not need to do anything
+            # because it cannot be a compound key.
+            if isinstance(k, str) and "." in k:
+                # Someone is using a more human-readable dataId
+                dimension, record = k.split(".", 1)
+                byRecord[dimension][record] = v
+            else:
+                newDataId[k] = v
+
+        if byRecord:
+            # Some record specifiers were found so we need to convert
+            # them to the Id form
+            for dimension, values in byRecord.items():
+                if dimension in newDataId:
+                    log.warning("DataId specified explicit %s dimension value of %s in addition to"
+                                " general record specifiers for it of %s.  Ignoring record information.",
+                                dimension, newDataId[dimension], str(values))
+                    continue
+
+                # Build up a WHERE expression -- use single quotes
+                def quote(s):
+                    if isinstance(s, str):
+                        return f"'{s}'"
```

Comment:
I'm hoping people are not going to be forming dataIds where they are doing `exposure.dattime_begin = ...` but maybe they will. It is *unique` so it would work...

Reply:
Heh, I guess we know that wouldn't work right now anyway, because none of the timespan fields do.